clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/deploy/ops/CopyLastAsgAtomicOperation.java

```diff
@@ -0,0 +1,552 @@
+import com.netflix.spinnaker.clouddriver.aws.model.SubnetData;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.NetflixAmazonCredentials;
+import com.netflix.spinnaker.clouddriver.aws.services.RegionScopedProviderFactory;
+import com.netflix.spinnaker.clouddriver.aws.services.RegionScopedProviderFactory.RegionScopedProvider;
+import com.netflix.spinnaker.clouddriver.aws.services.SecurityGroupService;
+import com.netflix.spinnaker.clouddriver.data.task.Task;
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository;
+import com.netflix.spinnaker.clouddriver.deploy.DeploymentResult;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidationErrors;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidationException;
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation;
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider;
+import java.util.*;
+import java.util.stream.Collectors;
+import jdk.internal.joptsimple.internal.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class CopyLastAsgAtomicOperation implements AtomicOperation<DeploymentResult> {
+  private static final Logger log = LoggerFactory.getLogger(CopyLastAsgAtomicOperation.class);
+  private static final String BASE_PHASE = "COPY_LAST_ASG";
+
+  private static Task getTask() {
+    return TaskRepository.threadLocalTask.get();
+  }
+
+  @Autowired private BasicAmazonDeployHandler basicAmazonDeployHandler;
+
+  @Autowired private BasicAmazonDeployDescriptionValidator basicAmazonDeployDescriptionValidator;
+
+  @Autowired private AmazonClientProvider amazonClientProvider;
+
+  @Autowired private AccountCredentialsProvider accountCredentialsProvider;
+
+  @Autowired private RegionScopedProviderFactory regionScopedProviderFactory;
+
+  @Autowired private LocalFileUserDataProperties localFileUserDataProperties;
+
+  private final BasicAmazonDeployDescription description;
+
+  public CopyLastAsgAtomicOperation(BasicAmazonDeployDescription description) {
+    this.description = description;
+  }
+
+  @Override
+  public DeploymentResult operate(List priorOutputs) {
+    getTask().updateStatus(BASE_PHASE, "Initializing Copy Last ASG Operation...");
+
+    final BasicAmazonDeployDescription.Source source = description.getSource();
+    Set<String> targetRegions =
+        Optional.ofNullable(description.getAvailabilityZones()).orElse(new HashMap<>()).keySet();
+    if (targetRegions.isEmpty()) {
+      targetRegions = new HashSet<>(Collections.singletonList(source.getRegion()));
+    }
+
+    if (!Strings.isNullOrEmpty(source.getRegion())
+        && !Strings.isNullOrEmpty(source.getAccount())
+        && !Strings.isNullOrEmpty(source.getAsgName())) {
```

Comment:
nit: This conditional is reoccurring in this method. Maybe worth pulling out into a helper method so conditionals are a little more readable?

`private static boolean hasAccountRegionAndAsg(String account, String region, String asgName)`?

Reply:
Will lift that out into a function.